### PR TITLE
Do not infer `yield*` type from contextual `TReturn`

### LINKED ISSUE #57903: Improper Type Inference in Async Generator Callback
### 🔎 Search Terms

async generator inference, async generator 

### 🕗 Version & Regression Information

- This is a crash
- This changed between versions 4.0.5 and 5.4.3
- This changed in commit or PR _______
- This is the behavior in every version I tried, and I reviewed the FAQ for entries about _________
- I was unable to test this on prior versions because _______

### ⏯ Playground Link

https://www.typescriptlang.org/play?target=99&jsx=0&ts=5.4.3#code/JYOwLgpgTgZghgYwgAgEoQM4FcA2YA8AKgDTICiAfMgN4BQyyA2gMoCeAtgEYD2OAdMEhQ4YblAC6ACgCUALmQBxCCGgix+MqUIVaAX1q0wrAA4oActzDpseAPIwiVALzJCyCAA9IIACYY0mLgEcCCspCGsVAD8yCoAbtDI8oQA3IYmKGRQUIQZ9o7ILm6e3n4BNgRYIADWINwA7iCkoDCJlMgxZEmxEAlQaUamyLbVuab52oWu7l7KZdZB+C2JqKRVtQ0g0Wjd8dADGcgAghisIAgAUtycRKTtLjKFVCdnCEoqwqJQGloUabQ+CAIHBwKAoBDcEAYMDIOCnc4AEW48lu5AokgAVtd5C9zlcbiQ0dInsgAApQbjsYAYCD4BZ4fD0YajPIOSYAH2QFisgTsbIoxCZXU5WRyrMctAofwBQJBYOQEKhMPYcGMopRmmQADF0SrjKYoPJJNAKYbyMSnFQtRaqI5JGCKvJ6QRCZQbeVFoSdf9BigAELcbjVKbUZDAHzyaFQUAAcxSyDAghwEEjYGjIDjsKwYAAFmIAJIR5BR2PIXQHIZHbN5qAhsNFksZ+MgODsFPFtOl8vpIYBoMAdUEOarubEUz7wYAZDQs6OzSOa2X-oDgaDwZDochWmAEDmJ0bw6n0zH3eTKdTac78BPSAAiMy2QgAfS1tgAqmYEbepcvZWuFRuMLbruC5iAeDadhmp4UlSNJ0ryBCgVAd4Ps+r4fl+P4GJ4xhiDCiqbsBe6BtUg65khACMhRMpIPBBoWR6xjaTIMHCrxIpIbHnFuVQIImkIAFTII8dAMGJAFKsgdHVPIE5TKwwAQDgPhCXA9RwIIW4QDuxFBrRJGFtILHicgfC5sokh6qKkiPJayD3o+L7vp+T5+rYtgANK3rC-gEWA0hGcZDB+bONY4tWY4uApSkqbC6maURSH6UGfBwBFUCGUFYlmTmFlWdkNnug5aHOQiT5HG+hAABK2Kg3lwhJ0IBQYJnIGCYBYFAIAznwvXSeE6Vlj5yATmRw7pWk4m6NI-w4XhjVAdpu6jUOSEAEzUQwyXVAxHbHnIZIwRe8EVNeJFjUhKGOehLluZ53mcsVTkYeVlU1XVUpPMZXEIBxP08ec-EgEJIlZSF0mySR8mKcpqnxYtOkTttmWtQwOV5aq1m2VQT03WVd1ecNfnNWDgGhWBxyDVFMOxWpGkIyB6Xbal6Uo6j6MgJZmMFdj9moc9LkVdVtX1b5gEk617Wdd1oa9Xw-Xk7WujDSt5ETcZ01pEAA

### 💻 Code

```ts
interface Result<T, E> {
  [Symbol.iterator](): Generator<E, T>
}

type NotResultOf<T> = T extends Result<any, any> ? never : T;
type ErrTypeOf<T> = T extends Result<unknown, infer E> ? E : never;
type OkTypeOf<T> = T extends Result<infer R, unknown> ? R : never;
type AsyncJob<T, E> = () => AsyncGenerator<E, T>;

declare const asyncDo: <T, E>(job: AsyncJob<T, E>) => Promise<Result<
  OkTypeOf<T> | NotResultOf<T>,
  E | ErrTypeOf<T>
>>;
declare const mapErr: <E, F>(mapper: (error: E) => F) => <T>(result: Result<T, E>) => Result<T, F>;

type Book = { id: string; title: string; authorId: string };
type Author = { id: string; name: string };
type BookWithAuthor = Book & { author: Author };

declare const fetchBook: (id: string) => Promise<Result<Book, "NOT_FOUND">>;

declare const fetchAuthor: (id: string) => Promise<Result<Author, "NOT_FOUND">>;

export const fetchBookWithAuthor1 =
  (bookId: string) =>
    asyncDo(async function* () {
      const book: Book = yield* await fetchBook(bookId)
        .then(mapErr(() => "NOT_FOUND_BOOK" as const))

      const author: Author = yield* await fetchAuthor(book.authorId)
        .then(mapErr(() => "NOT_FOUND_AUTHOR" as const))

      return { ...book, author } as BookWithAuthor;
    });

export const fetchBookWithAuthor2 =
  (bookId: string): Promise<Result<BookWithAuthor, "NOT_FOUND_BOOK" | "NOT_FOUND_AUTHOR">> =>
    asyncDo(async function* () {
      const book: Book = yield* await fetchBook(bookId)
        .then(mapErr(() => "NOT_FOUND_BOOK" as const))

      const author: Author = yield* await fetchAuthor(book.authorId)
        .then(mapErr(() => "NOT_FOUND_AUTHOR" as const))

      return { ...book, author } as BookWithAuthor;
    });
```

### 🙁 Actual behavior

Compilation Error in the function `fetchBookWithAuthor2`
> Type 'Author | BookWithAuthor' is not assignable to type 'Author'.
  Property 'name' is missing in type 'BookWithAuthor' but required in type 'Author'.(2322)
input.ts(17, 29): 'name' is declared here.

### 🙂 Expected behavior

No errors.
compilation result must be the same as for `fetchBookWithAuthor1`.

The only difference between two versions of the function is a return type specified.

The defenition file is emitted correctly:
```ts
interface Result<T, E> {
    [Symbol.iterator](): Generator<E, T>;
}
type Book = {
    id: string;
    title: string;
    authorId: string;
};
type Author = {
    id: string;
    name: string;
};
type BookWithAuthor = Book & {
    author: Author;
};
export declare const fetchBookWithAuthor1: (bookId: string) => Promise<Result<BookWithAuthor, "NOT_FOUND_BOOK" | "NOT_FOUND_AUTHOR">>;
export declare const fetchBookWithAuthor2: (bookId: string) => Promise<Result<BookWithAuthor, "NOT_FOUND_BOOK" | "NOT_FOUND_AUTHOR">>;
export {};
```

### Additional information about the issue

_No response_

### PR DESCRIPTION
**Pull request by rbuckton**
This changes the contextual type we produce for the operand to `yield*` so that we do not consider the `TReturn` type of the containing generator's return type as a source of inference since the `TReturn` of a generator has no bearing on the result of `yield*`. Instead, we substitute the contextual type for the entire `yield*` expression, e.g.:

```ts
interface Point {
  x: number;
  y: number;
}

declare function* g<U, UReturn>(
  getYield: () => U | undefined,
  getResult: () => UReturn,
): Generator<U, UReturn, unknown>;

async function* f(): AsyncGenerator</*T*/ number, /*TReturn*/ string, unknown> {

  const v1 = yield* g(
    () => /* a */,
    () => /* b */
  );
  
  const v2: Point = yield* g(
    () => /* c */,
    () => /* d */
  );
  
  return "done";
}
```

At (a), `U` should be inferred from `T` (i.e., `number`) since `f` will delegate `yield` from `g`.

At (b), `UReturn` should **not** be inferred from `TReturn` (i.e., `string`) since `f` does not delegate the return value of `g`. Since `v1` is untyped, `UReturn` is inferred as `unknown`.

At (c), `U` should be inferred from `T` (i.e., `number`) since `f` will delegate `yield` from `g`.

At (d), `UReturn` should **not** be inferred from `TReturn` (i.e., `string`) since `f` does not delegate the return value of `g`. Since `v2` has the type `Point`, `UReturn` is inferred as `Point`.

Fixes #57903

### CODE REVIEW

**Comment by rbuckton** [comment]
@typescript-bot perf test
@typescript-bot run dt
@typescript-bot test top400
@typescript-bot test tsserver top100
@typescript-bot user test this
@typescript-bot user test tsserver

**Comment by rbuckton** [comment]
> @rbuckton Here are the results of running the top 200 repos comparing `main` and `refs/pull/58621/merge`:
> 
> Something interesting changed - please have a look.

It doesn't seem like the crash is related? If I run the repro locally I have the same result with or without this change. I'll try running it again in case it was a quirk.

@typescript-bot test tsserver top100

**Comment by rbuckton** [comment]
Temporarily reverting change to validate crash.

@typescript-bot test tsserver top100
@typescript-bot user test tsserver

**Comment by jakebailey** [comment]
The server tests have issues; I don't think they properly detect old/new crashes like the tsc tests yet.

**Comment by rbuckton** [comment]
The crashes seem to be unrelated as they're still happening even after reverting the change.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -30839,9 +30839,19 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 if (!node.asteriskToken && contextualReturnType.flags & TypeFlags.Union) {
                     contextualReturnType = filterType(contextualReturnType, type => !!getIterationTypeOfGeneratorFunctionReturnType(IterationTypeKind.Return, type, isAsyncGenerator));
                 }
-                return node.asteriskToken
-                    ? contextualReturnType
-                    : getIterationTypeOfGeneratorFunctionReturnType(IterationTypeKind.Yield, contextualReturnType, isAsyncGenerator);
+                if (node.asteriskToken) {
+                    const iterationTypes = getIterationTypesOfGeneratorFunctionReturnType(contextualReturnType, isAsyncGenerator);
+                    const yieldType = iterationTypes?.yieldType ?? silentNeverType;
```
Comment (gabritto) [inline src/compiler/checker.ts:30844]: Why do we fall back to `silentNeverType`?

Comment (rbuckton) [reply src/compiler/checker.ts:30844]: We need to fall back to something we do not infer from, and a generator that doesn't `yield` infers a `T` of `never`. `silentNeverType` is non-inferable, so it seemed like the correct choice.

**Review by gabritto — APPROVED** [review]

**Pull request merged**